python/ray/autoscaler/v2/instance_manager/subscribers/threaded_ray_installer.py

```diff
@@ -31,65 +46,50 @@ def __init__(
             max_workers=self._max_concurrent_installs
         )
 
     def notify(self, events: List[InstanceUpdateEvent]) -> None:
         for event in events:
-            if event.new_instance_status == Instance.ALLOCATED:
+            if event.new_instance_status == Instance.RAY_INSTALLING:
                 self._install_ray_on_new_nodes(event.instance_id)
 
     def _install_ray_on_new_nodes(self, instance_id: str) -> None:
         allocated_instance, _ = self._instance_storage.get_instances(
             instance_ids={instance_id},
-            status_filter={Instance.ALLOCATED},
+            status_filter={Instance.RAY_INSTALLING},
         )
         for instance in allocated_instance.values():
+            assert instance.node_kind == NodeKind.WORKER
             self._ray_installation_executor.submit(
                 self._install_ray_on_single_node, instance
             )
 
     def _install_ray_on_single_node(self, instance: Instance) -> None:
-        assert instance.status == Instance.ALLOCATED
-        success, version = self._instance_storage.upsert_instance(
-            instance, expected_instance_version=instance.version
-        )
-        if not success:
-            logger.warning(
-                f"Failed to update instance {instance.instance_id} to RAY_INSTALLING"
-            )
-            # Do not need to handle failures, it will be covered by
-            # garbage collection.
-            return
+        assert instance.status == Instance.RAY_INSTALLING
 
         # install with exponential backoff
-        installed = False
         backoff_factor = 1
+        last_exception = None
         for _ in range(self._max_install_attempts):
-            installed = self._ray_installer.install_ray(instance, self._head_node_ip)
-            if installed:
-                break
+            try:
+                self._ray_installer.install_ray(instance, self._head_node_ip)
+                return
+            except Exception as e:
+                logger.info(
+                    f"Ray installation failed on instance {instance.cloud_instance_id}: {e}"
+                )
+                last_exception = e
+
             logger.warning("Failed to install ray, retrying...")
             time.sleep(self._install_retry_interval * backoff_factor)
             backoff_factor *= 2
 
-        if not installed:
-            instance.status = Instance.RAY_INSTALL_FAILED
-            success, version = self._instance_storage.upsert_instance(
-                instance,
-                expected_instance_version=version,
-            )
-        else:
-            instance.status = Instance.RAY_RUNNING
-            success, version = self._instance_storage.upsert_instance(
-                instance,
-                expected_instance_version=version,
+        self._error_queue.put_nowait(
+            RayInstallError(
+                im_instance_id=instance.instance_id,
+                details=str(last_exception),
```

Comment:
Put the last retry error in the queue and let the autoscaler v2 reconciler handle the error transition.